### PR TITLE
SMT checker for various things

### PR DESCRIPTION
**Pull request by chriseth**
This can now handle conditions and while loops. It checks for conditions being constant and correctly erases knowledge about local variables at control flow join paths.

### CODE REVIEW

**Comment by chriseth** [comment]
This can be merged in the current state.

**Comment by axic** [comment]
Consider merging #3030 and rebase this after.

**Comment by chriseth** [comment]
Is #3030 in conflict with this?

File: libsolidity/ast/ASTForward.h [hunk 1]
```diff
@@ -57,6 +57,7 @@ class UserDefinedTypeName;
 class FunctionTypeName;
 class Mapping;
 class ArrayTypeName;
+class InlineAssembly;
```
Comment (axic) [inline libsolidity/ast/ASTForward.h]: It doesn't seem to be used in this PR, but it was missing anyway.

**Comment by axic** [comment]
Should this be reviewed/merged or #3032?

**Comment by chriseth** [comment]
#3032 is an extension of this. Please review this one.

**Comment by axic** [comment]
This is failing tests though:
```
/home/travis/build/ethereum/solidity/libsolidity/formal/SMTChecker.cpp(674): fatal error in "void dev::solidity::SMTChecker::setZeroValue(const dev::solidity::Declaration &)": std::exception: 
/home/travis/build/ethereum/solidity/test/libsolidity/SMTChecker.cpp(91): last checkpoint
```

**Comment by chriseth** [comment]
Cherry-picked a commit from #3032 that should fix the tests.

**Comment by chriseth** [comment]
The remaining failure seems to be a travis issue.

File: libsolidity/ast/ASTForward.h (hunk 1 above)
Comment (jackpot1136) [reply libsolidity/ast/ASTForward.h]: Who was assigned to me and to teach me and help me out

**Comment by axic** [comment]
Any reasonable way to split this? It seems to do a lot of different improvements.

**Comment by chriseth** [comment]
Can you review the individual commits? "track variables" is probably the largest. I can try to split it up, but I fear it will be a lot of work.

**Comment by axic** [comment]
I think I can pull out the renaming and the exceptions easily, the rest is mushy to me at the moment at least.

Based on the tests it seems it supports at least more variables, branching, bools.

File: libsolidity/formal/Z3Interface.cpp [hunk 2]
```diff
@@ -118,8 +118,7 @@ z3::expr Z3Interface::toZ3Expr(Expression const& _expr)
 		{">=", 2},
 		{"+", 2},
 		{"-", 2},
-		{"*", 2},
-		{">=", 2}
```
Comment (axic) [inline libsolidity/formal/Z3Interface.cpp]: Ah, `>=` was duplicated.

**Comment by axic** [comment]
Rebased on #3091.

**Comment by axic** [comment]
@chriseth you've pushed an old version over the rebased one. Any reason for it? In any case, please rebase.

**Comment by chriseth** [comment]
You mean I rebased on top of an old version? Sorry, I don't know why that happened. I hope I did not lose any commits. Fixing...

**Comment by axic** [comment]
I've rebased it on top of the cleanup branch, but hasn't changed anything. It should be safe as long as you had the last version before that :)

**Comment by axic** [comment]
Are there are any asserts which should be `solUnimplementedAssert`?

**Comment by chriseth** [comment]
I will take a look at the next iteration.

**Comment by chriseth** [comment]
@axic can this be merged? We already have two other finished pull requests waiting on top of this one.

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
Rebased before merging as it was fairly old (had circleci too).

**Pull request merged**